src/mailbox.rs

```diff
@@ -78,8 +100,30 @@ impl MailboxService for MailboxGrpc {
         } else {
             return Err(tonic::Status::new(tonic::Code::InvalidArgument, "must specify uuid"));
         };
+        let public_key = if request.public_key.is_empty() {
+            None
+        } else {
+            Some(base64::encode(&request.public_key))
+        };
+
+        if self.config.user_auth_enabled {
+            let public_key = public_key.clone().ok_or(Status::permission_denied("auth is enabled, but a public_key wasn't sent"))?;
+            let cache = self.cache.lock().unwrap();
+
+            match cache.get_public_key_state(&public_key) {
+                PublicKeyState::Local => {
+                    if let Some(cached_key) = cache.public_keys.get(&public_key) {
+                        cached_key.auth()?.authorize(&metadata)
+                    } else {
+                        Err(Status::internal("this should not happen - key was resolved to Local state and then could not be fetched"))
+                    }
+                },
+                PublicKeyState::Remote => Err(Status::permission_denied(format!("remote public key {} - use the replicate route", &public_key))),
```

Comment:
Should this match the other error messages you changed? i.e. something like `remote public key {} - fetch on its own instance`... though maybe `ack on its own instance`?

Reply:
👍